samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,188 @@
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def local_path(self) -> Optional[Path]:
+        return self._local_path
+
+    @local_path.setter
+    def local_path(self, value: Path) -> None:
+        self._local_path = value
+
+    @property
+    def clone_attempted(self) -> bool:
+        return self._clone_attempted
+
+    @clone_attempted.setter
+    def clone_attempted(self, value: bool) -> None:
+        self._clone_attempted = value
+
+    @property
+    def auto_clone(self) -> bool:
+        return self._auto_clone
+
+    @staticmethod
+    def _ensure_clone_directory_exists(clone_dir: Path) -> None:
+        try:
+            clone_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+        except OSError as ex:
+            LOG.warning("WARN: Unable to create clone directory.", exc_info=ex)
+            raise
+
+    @staticmethod
+    def _git_executable() -> str:
+        if platform.system().lower() == "windows":
+            executables = ["git", "git.cmd", "git.exe", "git.bat"]
+        else:
+            executables = ["git"]
+
+        for executable in executables:
+            try:
+                subprocess.Popen([executable], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                # No exception. Let's pick this
+                return executable
+            except OSError as ex:
+                LOG.debug("Unable to find executable %s", executable, exc_info=ex)
+
+        raise OSError("Cannot find git, was looking at executables: {}".format(executables))
+
+    def clone(self, clone_dir: Path, clone_name: Optional[str] = None, replace_existing: bool = False) -> Path:
+        """
+        creates a local clone of this Git repository
+
+        Parameters
+        ----------
+        clone_dir: Path
+            The directory to create the local clone inside
+        clone_name: Optional[str]
+            The dirname of the local clone, if not provided, repository's name (self._name) will be used
+        replace_existing: bool
+            Whether to replace the current local clone directory if already exists or not
+
+        Returns
+        -------
+            The path of the created local clone
+
+        Raises
+        ------
+        OSError:
+            when file management errors like unable to mkdir, copytree, rmtree ...etc
+        CloneRepoException:
+            General errors like for example; if an error occurred while running `git clone`
+            or if the local_clone already exists and replace_existing is not set
+        CloneRepoUnstableStateException:
+            when reaching unstable state, for example with replace_existing flag set, unstable state can happen
+            if removed the current local clone but failed to copy the new one from the temp location to the destination
+        """
+        if not self._auto_clone:  # Unit test escape hatch
```

Comment:
If I initialize the GitRepo with `auto_clone` at `False`, I can never clone?

Reply:
Not sure the `auto_clone` property is meant to be inside the class.
If so, I would expect it to control if the `__init__` method is launching a clone.

Reply:
As I understood from the code, `auto_clone` is just a flag used in the unit-testing to turn-off cloning. check in `tests/unit/commands/init/test_cli.py` for example [this line](https://github.com/aws/aws-sam-cli/blob/13a4250c137ce1f9e0087b7a17cca434b8c32bff/tests/unit/commands/init/test_cli.py#L66)  

in reality the `cli` method of `sam init` always call `do_cli` with `auto_clone` set to true see [here](https://github.com/aws/aws-sam-cli/blob/c99f1eee1ef4b23ffffde85e6c5f669b3970f083/samcli/commands/init/__init__.py#L259)